pennylane/ops/qubit/observables.py

```diff
@@ -528,9 +533,9 @@ def compute_eigvals(basis_state):  # pylint: disable=arguments-differ
         >>> BasisStateProjector.compute_eigvals([0, 1])
         [0. 1. 0. 0.]
         """
-        w = np.zeros(2 ** len(basis_state))
-        idx = int("".join(str(i) for i in basis_state), 2)
-        w[idx] = 1
+        w = qml.math.zeros(2 ** len(basis_state), like=basis_state)
+        idx = qml.math.dot(basis_state, 2 ** qml.math.arange(len(basis_state) - 1, -1, -1))
+        w = qml.math.where(w, idx, 1)
```

Comment:
```suggestion
        qml.math.put(w, idx, 1)
```
The `where` function doesn't do the same thing as updating the value at a particular index. Instead, it looks at the first array (for example [0, 1]), and gives you `idx` for each element that is True, and 1 for each element that is False. So, `qml.math.where([0, 1, 0], 2, 1)` gives you `array([1, 2, 1])`. 

Changing it to `put` should fix quite a few of the errors you are seeing. It will update the index (and return None), just like updating the index as `w[idx]` did before, but in a way that is compatible with more interfaces. 

I'd recommend trying this change, and then taking a look at what errors remain and seeing if it's clearer to make sense of!

Reply:
thanks for the explanation. 👍